
```scala
package ohnosequences.mg7.loquats

import ohnosequences.mg7._

import ohnosequences.mg7.bio4j._, taxonomyTree._, titanTaxonomyTree._

import ohnosequences.loquat._

import ohnosequences.statika._
import ohnosequences.cosas._, types._, klists._
import ohnosequences.datasets._
import ohnosequences.fastarious._, fasta._, fastq._

import better.files._
import com.github.tototoshi.csv._

import com.bio4j.titan.model.ncbiTaxonomy.TitanNCBITaxonomyGraph


case object statsDataProcessing extends DataProcessingBundle()(
  input = data.statsInput,
  output = data.statsOutput
) {

  def countReads(
    parser: Iterator[String] => Iterator[Any],
    file: File
  ): Integer = {
    val source = io.Source.fromFile( file.toJava )
    val readsNumber = parser( source.getLines ).length
    source.close()
    readsNumber
  }


  def instructions: ohnosequences.statika.AnyInstructions = say("Running stats loquat")

  def process(context: ProcessingContext[Input]): AnyInstructions { type Out <: OutputFiles } = {

    val statsCSV: File = (context / "output" / "stats.csv").createIfNotExists()
    val sampleID: String = context.inputFile(data.sampleID).contentAsString

    val reads1gz: File = context.inputFile(data.pairedReads1)
    val reads1fastq: File = File(reads1gz.path.toString.stripSuffix(".gz"))

    cmd("gunzip")(reads1gz.path.toString) -&-
    LazyTry {
      val csvWriter = csv.newWriter(statsCSV)

      // header:
      csvWriter.writeRow(csv.statsHeader)

      // values:
      // NOTE: careful, the order has to coincide with the header
      // TODO: use csv.Row here
      val stats: Seq[String] = Seq(
        sampleID,

        countReads( parseFastqDropErrors, reads1fastq ).toString,
        countReads( parseFastqDropErrors, context.inputFile(data.mergedReads) ).toString,
        countReads( parseFastqDropErrors, context.inputFile(data.pair1NotMerged) ).toString,

        countReads( parseFastaDropErrors, context.inputFile(data.blastNoHits) ).toString
      )

      csvWriter.writeRow(stats)

      csvWriter.close()
    } -&-
    success(s"Stats for the [${sampleID}] are ready",
      data.sampleStatsCSV(statsCSV) ::
      *[AnyDenotation { type Value <: FileResource }]
    )
  }

}

```




[main/scala/mg7/bio4j/bundle.scala]: ../bio4j/bundle.scala.md
[main/scala/mg7/bio4j/taxonomyTree.scala]: ../bio4j/taxonomyTree.scala.md
[main/scala/mg7/bio4j/titanTaxonomyTree.scala]: ../bio4j/titanTaxonomyTree.scala.md
[main/scala/mg7/csv.scala]: ../csv.scala.md
[main/scala/mg7/data.scala]: ../data.scala.md
[main/scala/mg7/dataflow.scala]: ../dataflow.scala.md
[main/scala/mg7/dataflows/full.scala]: ../dataflows/full.scala.md
[main/scala/mg7/dataflows/noFlash.scala]: ../dataflows/noFlash.scala.md
[main/scala/mg7/loquats/1.flash.scala]: 1.flash.scala.md
[main/scala/mg7/loquats/2.split.scala]: 2.split.scala.md
[main/scala/mg7/loquats/3.blast.scala]: 3.blast.scala.md
[main/scala/mg7/loquats/4.assign.scala]: 4.assign.scala.md
[main/scala/mg7/loquats/5.merge.scala]: 5.merge.scala.md
[main/scala/mg7/loquats/6.count.scala]: 6.count.scala.md
[main/scala/mg7/loquats/7.stats.scala]: 7.stats.scala.md
[main/scala/mg7/loquats/8.summary.scala]: 8.summary.scala.md
[main/scala/mg7/package.scala]: ../package.scala.md
[main/scala/mg7/parameters.scala]: ../parameters.scala.md
[test/scala/mg7/counts.scala]: ../../../../test/scala/mg7/counts.scala.md
[test/scala/mg7/lca.scala]: ../../../../test/scala/mg7/lca.scala.md
[test/scala/mg7/pipeline.scala]: ../../../../test/scala/mg7/pipeline.scala.md
[test/scala/mg7/taxonomy.scala]: ../../../../test/scala/mg7/taxonomy.scala.md