
```scala
package ohnosequences.mg7.loquats

import ohnosequences.mg7._

import ohnosequences.loquat._

import ohnosequences.statika._

import ohnosequences.{ flash => f }, f.api._

import ohnosequences.cosas._, types._, klists._

import ohnosequences.datasets._

import better.files._


case class flashDataProcessing[MD <: AnyMG7Parameters](val md: MD)
extends DataProcessingBundle(
  bundles.flash
)(
  input = data.flashInput,
  output = data.flashOutput
) {

  def instructions: AnyInstructions = say("I'll be fast as a flash!")

  // TODO FLASh stuff change options, derived from reads type
  def process(context: ProcessingContext[Input]): AnyInstructions { type Out <: OutputFiles } = {

    val reads1gz: File = context.inputFile(data.pairedReads1)
    val reads2gz: File = context.inputFile(data.pairedReads2)

    val reads1fastq: File = File(reads1gz.path.toString.stripSuffix(".gz"))
    val reads2fastq: File = File(reads2gz.path.toString.stripSuffix(".gz"))

    // define input
    lazy val flashInput = FlashInputAt(
      reads1fastq,
      reads2fastq
    )

    // define output
    lazy val flashOutput = FlashOutputAt((context / "output"), prefix = "")

    // the FLASh cmd we are going to run
    lazy val flashExpr = FlashExpression(
      flash.arguments(
        f.api.input(flashInput)   ::
        f.api.output(flashOutput) ::
        *[AnyDenotation]
      ),
      md.flashOptions
    )

    // run expression, hope for the best
    cmd("gunzip")(reads1gz.path.toString) -&-
    cmd("gunzip")(reads2gz.path.toString) -&-
    seqToInstructions(flashExpr.cmd) -&-
    success(
      "FLASh merged reads, much success so fast",
      data.mergedReads(flashOutput.mergedReads) ::
      data.flashStats(flashOutput.lengthNumericHistogram) ::
      *[AnyDenotation { type Value <: FileResource }]
    )
  }
}

```




[main/scala/metagenomica/bio4j/taxonomyTree.scala]: ../bio4j/taxonomyTree.scala.md
[main/scala/metagenomica/bio4j/titanTaxonomyTree.scala]: ../bio4j/titanTaxonomyTree.scala.md
[main/scala/metagenomica/bundles/bio4jTaxonomy.scala]: ../bundles/bio4jTaxonomy.scala.md
[main/scala/metagenomica/bundles/blast.scala]: ../bundles/blast.scala.md
[main/scala/metagenomica/bundles/blast16s.scala]: ../bundles/blast16s.scala.md
[main/scala/metagenomica/bundles/flash.scala]: ../bundles/flash.scala.md
[main/scala/metagenomica/bundles/gis.scala]: ../bundles/gis.scala.md
[main/scala/metagenomica/data.scala]: ../data.scala.md
[main/scala/metagenomica/dataflows/standard.scala]: ../dataflows/standard.scala.md
[main/scala/metagenomica/loquats/1.flash.scala]: 1.flash.scala.md
[main/scala/metagenomica/loquats/2.split.scala]: 2.split.scala.md
[main/scala/metagenomica/loquats/3.blast.scala]: 3.blast.scala.md
[main/scala/metagenomica/loquats/4.merge.scala]: 4.merge.scala.md
[main/scala/metagenomica/loquats/5.assignment.scala]: 5.assignment.scala.md
[main/scala/metagenomica/loquats/6.counting.scala]: 6.counting.scala.md
[main/scala/metagenomica/package.scala]: ../package.scala.md
[main/scala/metagenomica/parameters.scala]: ../parameters.scala.md
[test/scala/bundles.scala]: ../../../../test/scala/bundles.scala.md
[test/scala/lca.scala]: ../../../../test/scala/lca.scala.md
[test/scala/metagenomica/pipeline.scala]: ../../../../test/scala/metagenomica/pipeline.scala.md