
```scala
package ohnosequences.mg7.loquats

import ohnosequences.mg7._
import ohnosequences.loquat._
import ohnosequences.statika._
import ohnosequences.cosas._, types._, klists._
import ohnosequences.datasets._
import ohnosequences.fastarious._, fasta._

import better.files._
import collection.JavaConversions._


case class splitDataProcessing(params: AnyMG7Parameters) extends DataProcessingBundle()(
  input = data.splitInput,
  output = data.splitOutput
) {

  def instructions: AnyInstructions = say("Splitting, cutting, separating")

  def process(context: ProcessingContext[Input]): AnyInstructions { type Out <: OutputFiles } = {

    val outputDir = context / "chunks"

    LazyTry {
      outputDir.createDirectories()

      val lines: Iterator[String] = context.inputFile(data.mergedReads).lines

      lazy val fastas: Iterator[String] = params.splitInputFormat match {
        // if input is FastQ, each read is just 4 lines
        case FastQInput => lines.grouped(4).map { quartet =>
          FASTA(
            fasta.header(FastaHeader(quartet(0))) ::
            fasta.sequence(FastaSequence(quartet(1))) ::
            *[AnyDenotation]
          ).toLines
        }
        // if it's Fasta, we parse it, group and discard unparsed pieces
        case FastaInput => fasta.parseMapFromLines(lines).map { fastaMap =>
          FASTA(
            fasta.header(FastaHeader(fastaMap(fasta.header.label))) ::
            fasta.sequence(FastaSequence(fastaMap(fasta.sequence.label))) ::
            *[AnyDenotation]
          ).toLines
        }
      }

      fastas
        .grouped(params.splitChunkSize)
        .zipWithIndex
        .foreach { case (chunk, n) =>

          (outputDir / s"chunk.${n}.fasta")
            .overwrite(chunk.mkString("\n"))
        }
    } -&-
    success("chunk-chunk-chunk!",
      data.fastaChunks(outputDir) ::
      *[AnyDenotation { type Value <: FileResource }]
    )

  }
}

```




[main/scala/metagenomica/bio4j/taxonomyTree.scala]: ../bio4j/taxonomyTree.scala.md
[main/scala/metagenomica/bio4j/titanTaxonomyTree.scala]: ../bio4j/titanTaxonomyTree.scala.md
[main/scala/metagenomica/bundles/bio4jTaxonomy.scala]: ../bundles/bio4jTaxonomy.scala.md
[main/scala/metagenomica/bundles/blast.scala]: ../bundles/blast.scala.md
[main/scala/metagenomica/bundles/filterGIs.scala]: ../bundles/filterGIs.scala.md
[main/scala/metagenomica/bundles/flash.scala]: ../bundles/flash.scala.md
[main/scala/metagenomica/bundles/referenceDB.scala]: ../bundles/referenceDB.scala.md
[main/scala/metagenomica/bundles/referenceMap.scala]: ../bundles/referenceMap.scala.md
[main/scala/metagenomica/data.scala]: ../data.scala.md
[main/scala/metagenomica/dataflow.scala]: ../dataflow.scala.md
[main/scala/metagenomica/dataflows/noFlash.scala]: ../dataflows/noFlash.scala.md
[main/scala/metagenomica/dataflows/standard.scala]: ../dataflows/standard.scala.md
[main/scala/metagenomica/loquats/1.flash.scala]: 1.flash.scala.md
[main/scala/metagenomica/loquats/2.split.scala]: 2.split.scala.md
[main/scala/metagenomica/loquats/3.blast.scala]: 3.blast.scala.md
[main/scala/metagenomica/loquats/4.merge.scala]: 4.merge.scala.md
[main/scala/metagenomica/loquats/5.assignment.scala]: 5.assignment.scala.md
[main/scala/metagenomica/loquats/6.counting.scala]: 6.counting.scala.md
[main/scala/metagenomica/package.scala]: ../package.scala.md
[main/scala/metagenomica/parameters.scala]: ../parameters.scala.md
[test/scala/bundles.scala]: ../../../../test/scala/bundles.scala.md
[test/scala/lca.scala]: ../../../../test/scala/lca.scala.md
[test/scala/metagenomica/pipeline.scala]: ../../../../test/scala/metagenomica/pipeline.scala.md