
```scala
package ohnosequences.mg7.loquats

import ohnosequences.mg7._

import ohnosequences.loquat._

import ohnosequences.statika._

import ohnosequences.blast.api._

import ohnosequences.cosas._, types._, klists._

import ohnosequences.datasets._

import ohnosequences.fastarious._, fasta._

import better.files._

import sys.process._


case class blastDataProcessing[MD <: AnyMG7Parameters](val md: MD)
extends DataProcessingBundle(
  bundles.blast,
  md.referenceDB
)(
  input = data.blastInput,
  output = data.blastOutput
) {

  def instructions: AnyInstructions = say("Let the blasting begin!")

  def process(context: ProcessingContext[Input]): AnyInstructions { type Out <: OutputFiles } = {

    val totalOutput = context / "blastAll.csv"

    LazyTry {
      // NOTE: once we update to better-files 2.15.+, use `file.lineIterator` here (it's autoclosing):
      val source = io.Source.fromFile( context.inputFile(data.fastaChunk).toJava )

      fasta.parseMapFromLines(source.getLines) foreach { fastaMap =>

        val read = FASTA(
          header(FastaHeader(fastaMap(fasta.header.label))) ::
          sequence(FastaSequence(fastaMap(fasta.sequence.label))) ::
          *[AnyDenotation]
        )

        val inFile = (context / "read.fa").overwrite(read.toLines.stripSuffix("\n"))
        val outFile = (context / "blastRead.csv").clear()

        val expr = blastn(
          outputRecord = md.blastOutRec,
          argumentValues =
            db(md.referenceDB.dbName) ::
            query(inFile) ::
            out(outFile) ::
            *[AnyDenotation],
          optionValues = md.blastOptions.value
        )
        println(expr.toSeq.mkString(" "))

        // BAM!!
        expr.toSeq.!!

        // append results for this read to the total output
        totalOutput.append(outFile.contentAsString)
      }

      // it's important to close the stream:
      source.close()
    } -&-
    success(
      "much blast, very success!",
      data.blastChunkOut(totalOutput) ::
      *[AnyDenotation { type Value <: FileResource }]
    )

  }
}

```




[main/scala/metagenomica/bio4j/taxonomyTree.scala]: ../bio4j/taxonomyTree.scala.md
[main/scala/metagenomica/bio4j/titanTaxonomyTree.scala]: ../bio4j/titanTaxonomyTree.scala.md
[main/scala/metagenomica/bundles/bio4jTaxonomy.scala]: ../bundles/bio4jTaxonomy.scala.md
[main/scala/metagenomica/bundles/blast.scala]: ../bundles/blast.scala.md
[main/scala/metagenomica/bundles/filterGIs.scala]: ../bundles/filterGIs.scala.md
[main/scala/metagenomica/bundles/flash.scala]: ../bundles/flash.scala.md
[main/scala/metagenomica/bundles/referenceDB.scala]: ../bundles/referenceDB.scala.md
[main/scala/metagenomica/bundles/referenceMap.scala]: ../bundles/referenceMap.scala.md
[main/scala/metagenomica/data.scala]: ../data.scala.md
[main/scala/metagenomica/dataflow.scala]: ../dataflow.scala.md
[main/scala/metagenomica/dataflows/noFlash.scala]: ../dataflows/noFlash.scala.md
[main/scala/metagenomica/dataflows/standard.scala]: ../dataflows/standard.scala.md
[main/scala/metagenomica/loquats/1.flash.scala]: 1.flash.scala.md
[main/scala/metagenomica/loquats/2.split.scala]: 2.split.scala.md
[main/scala/metagenomica/loquats/3.blast.scala]: 3.blast.scala.md
[main/scala/metagenomica/loquats/4.merge.scala]: 4.merge.scala.md
[main/scala/metagenomica/loquats/5.assignment.scala]: 5.assignment.scala.md
[main/scala/metagenomica/loquats/6.counting.scala]: 6.counting.scala.md
[main/scala/metagenomica/package.scala]: ../package.scala.md
[main/scala/metagenomica/parameters.scala]: ../parameters.scala.md
[test/scala/bundles.scala]: ../../../../test/scala/bundles.scala.md
[test/scala/lca.scala]: ../../../../test/scala/lca.scala.md
[test/scala/metagenomica/pipeline.scala]: ../../../../test/scala/metagenomica/pipeline.scala.md