
```scala
package ohnosequences.mg7.loquats

import ohnosequences.mg7._

import ohnosequences.loquat._

import ohnosequences.statika._

import ohnosequences.blast.api._

import ohnosequences.cosas._, types._, klists._

import ohnosequences.datasets._

import ohnosequences.fastarious._
import ohnosequences.fastarious.fasta._
import ohnosequences.fastarious.fastq._

import better.files._

import sys.process._


case class blastDataProcessing[MD <: AnyMG7Parameters](val md: MD)
extends DataProcessingBundle(
  bundles.blast,
  bundles.blast16s
)(
  input = data.blastInput,
  output = data.blastOutput
) {

  def instructions: AnyInstructions = say("Let the blasting begin!")

  def process(context: ProcessingContext[Input]): AnyInstructions { type Out <: OutputFiles } = {

    val totalOutput = context / "blastAll.csv"

    LazyTry {
      // NOTE: once we update to better-files 2.15.+, use `file.lineIterator` here (it's autoclosing):
      lazy val source = io.Source.fromFile( context.inputFile(data.readsChunk).toJava )

      source.getLines.grouped(4) foreach { quartet =>
        // println(quartet.mkString("\n"))

        // we only care about the id and the seq here
        val read = FASTA(
            header(FastqId(quartet(0)).toFastaHeader) ::
            fasta.sequence(FastaLines(quartet(1)))    ::
            *[AnyDenotation]
          )

        val readFile = context / "read.fa"
        readFile
          .createIfNotExists()
          .appendLines(read.toLines: _*)

        val outFile = context / "blastRead.csv"

        val expr = blastn(
          outputRecord = md.blastOutRec,
          argumentValues = blastn.arguments(
            db(bundles.blast16s.dbName) ::
            query(readFile) ::
            out(outFile) ::
            *[AnyDenotation]
          ).value,
          optionValues   = blastn.defaults.update(
            num_threads(1) ::
            word_size(42) ::
            max_target_seqs(10) ::
            evalue(0.001) ::
            blastn.task(blastn.megablast) ::
            *[AnyDenotation]
          ).value
        )
        println(expr.toSeq.mkString(" "))

        // BAM!!!
        val exitCode = expr.toSeq.!
        println(s"BLAST EXIT CODE: ${exitCode}")

        // we should have something in args getV out now. Append it!
        println(s"Appending [${outFile.path}] to [${totalOutput.path}]")
        totalOutput
          .createIfNotExists()
          .append(outFile.contentAsString)

        // clean up
        readFile.delete(true)
        outFile.delete(true)
      }

      // it's important to close the stream:
      source.close()
    } -&-
    success(
      "much blast, very success!",
      data.blastChunkOut(totalOutput) ::
      *[AnyDenotation { type Value <: FileResource }]
    )

  }
}

```




[main/scala/metagenomica/bio4j/taxonomyTree.scala]: ../bio4j/taxonomyTree.scala.md
[main/scala/metagenomica/bio4j/titanTaxonomyTree.scala]: ../bio4j/titanTaxonomyTree.scala.md
[main/scala/metagenomica/bundles/bio4jTaxonomy.scala]: ../bundles/bio4jTaxonomy.scala.md
[main/scala/metagenomica/bundles/blast.scala]: ../bundles/blast.scala.md
[main/scala/metagenomica/bundles/blast16s.scala]: ../bundles/blast16s.scala.md
[main/scala/metagenomica/bundles/flash.scala]: ../bundles/flash.scala.md
[main/scala/metagenomica/bundles/gis.scala]: ../bundles/gis.scala.md
[main/scala/metagenomica/data.scala]: ../data.scala.md
[main/scala/metagenomica/dataflows/standard.scala]: ../dataflows/standard.scala.md
[main/scala/metagenomica/loquats/1.flash.scala]: 1.flash.scala.md
[main/scala/metagenomica/loquats/2.split.scala]: 2.split.scala.md
[main/scala/metagenomica/loquats/3.blast.scala]: 3.blast.scala.md
[main/scala/metagenomica/loquats/4.merge.scala]: 4.merge.scala.md
[main/scala/metagenomica/loquats/5.assignment.scala]: 5.assignment.scala.md
[main/scala/metagenomica/loquats/6.counting.scala]: 6.counting.scala.md
[main/scala/metagenomica/package.scala]: ../package.scala.md
[main/scala/metagenomica/parameters.scala]: ../parameters.scala.md
[test/scala/bundles.scala]: ../../../../test/scala/bundles.scala.md
[test/scala/lca.scala]: ../../../../test/scala/lca.scala.md
[test/scala/metagenomica/pipeline.scala]: ../../../../test/scala/metagenomica/pipeline.scala.md